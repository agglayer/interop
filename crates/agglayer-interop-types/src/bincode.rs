pub use bincode::{Error, Options, Result};

/// Bincode configuration. Deliberately inaccessible from the outside.
mod options {
    use bincode::config::{
        BigEndian, Bounded, DefaultOptions as BincodeDefaultOptions, FixintEncoding, Options as _,
        WithOtherEndian, WithOtherIntEncoding, WithOtherLimit,
    };

    pub type Default =
        WithOtherIntEncoding<WithOtherEndian<BincodeDefaultOptions, BigEndian>, FixintEncoding>;

    #[inline]
    pub fn default() -> Default {
        bincode::options().with_big_endian().with_fixint_encoding()
    }

    pub type Limited<T> = WithOtherLimit<T, Bounded>;

    pub type SP1v4 = BincodeDefaultOptions;

    #[inline]
    pub fn sp1v4() -> SP1v4 {
        bincode::options()
    }
}

/// Bincode codec with opinionated settings.
#[derive(Clone, Debug)]
pub struct Codec<Opts>(Opts);

/// Create a bincode codec with default agglayer settings.
#[inline]
pub fn default() -> Codec<options::Default> {
    Codec(options::default())
}

/// Create a bincode codec with settings used by `sp1`.
#[inline]
pub fn sp1v4() -> Codec<options::SP1v4> {
    Codec(options::sp1v4())
}

impl<Opts: Options> Codec<Opts> {
    /// Impose a limit on encoding / decoding size.
    #[inline]
    pub fn with_limit(self, max: u64) -> Codec<options::Limited<Opts>> {
        Codec(self.0.with_limit(max))
    }

    /// Encode an object into a byte vector.
    #[inline]
    pub fn serialize<T>(self, item: &T) -> Result<Vec<u8>>
    where
        T: ?Sized + serde::Serialize,
    {
        self.0.serialize(item)
    }

    /// Encode an object into a writer.
    #[inline]
    pub fn serialize_into<W, T>(self, writer: W, item: &T) -> Result<()>
    where
        W: std::io::Write,
        T: ?Sized + serde::Serialize,
    {
        self.0.serialize_into(writer, item)
    }

    /// Decode an object from a slice.
    #[inline]
    pub fn deserialize<'a, T>(self, bytes: &'a [u8]) -> Result<T>
    where
        T: serde::Deserialize<'a>,
    {
        self.0.deserialize(bytes)
    }

    /// Decode an object from a reader.
    #[inline]
    pub fn deserialize_from<T, R>(self, reader: R) -> Result<T>
    where
        T: serde::de::DeserializeOwned,
        R: std::io::Read,
    {
        self.0.deserialize_from(reader)
    }
}
