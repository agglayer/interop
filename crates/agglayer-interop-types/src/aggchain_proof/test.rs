use super::*;

use agglayer_primitives::U256;

#[test]
fn regression_sp1_serialization_roundtrip_fail() {
    // Conclusion: sp1 serialization is not deterministic, removed the equality
    // check.
    use bincode::Options;
    let bytes = hex::decode("00000000000000000000000000000000000000000000fb00000100000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000").unwrap();
    let input: SP1StarkProof = bincode::options()
        .deserialize(&bytes)
        .expect("failed first deserialization, would be fine");
    let serialized: Vec<u8> = crate::bincode::default()
        .serialize(&input)
        .expect("failed serialization, unexpected");
    let _output: SP1StarkProof = crate::bincode::default()
        .deserialize(&serialized)
        .expect("failed second deserialization, is unexpected");
}

#[rstest::rstest]
#[case(
    "multisig-empty",
    AggchainData::MultisigOnly { multisig: MultisigPayload(vec![]) },
)]
#[case(
    "multisig-none",
    AggchainData::MultisigOnly { multisig: MultisigPayload(vec![None]) },
)]
#[case(
    "multisig-onesig",
    AggchainData::MultisigOnly { multisig: MultisigPayload(vec![
        Some(Signature::new(U256::ZERO, U256::ZERO, false)),
    ]) },
)]
#[case(
    "ecdsa",
    AggchainData::ECDSA { signature: Signature::new(U256::ZERO, U256::ZERO, false),
})]
fn aggchaindata_json(#[case] name: &str, #[case] aggchain_data: AggchainData) {
    #[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
    struct Container {
        #[serde(flatten)]
        aggchain_data: AggchainData,
    }

    let container = Container { aggchain_data };
    let json_str = serde_json::to_string_pretty(&container).expect("serialization failed");
    insta::assert_snapshot!(name, json_str);
}
