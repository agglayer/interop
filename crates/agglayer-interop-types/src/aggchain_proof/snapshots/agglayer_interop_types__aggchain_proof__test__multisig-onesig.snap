---
source: crates/agglayer-interop-types/src/aggchain_proof/test.rs
expression: json_str
---
{
  "multisig": [
    {
      "r": "0x0",
      "s": "0x0",
      "odd_y_parity": false
    }
  ]
}
